src/coreclr/jit/morph.cpp

```diff
@@ -12651,79 +12662,199 @@ GenTree* Compiler::fgOptimizeEqualityComparisonWithConst(GenTreeOp* cmp)
+                    // recognizes certain patterns that involve GT_NE (e.g (x & 4) != 0) and fails
+                    // if GT_GT is used instead. Transform (x GT_GT.unsigned 0) into (x GT_NE 0)
+                    // and (x GT_LE.unsigned 0) into (x GT_EQ 0). The later case is rare, it sometimes
+                    // occurs as a result of branch inversion.
+                    oper = (oper == GT_LE) ? GT_EQ : GT_NE;
+                    cmp->gtFlags &= ~GTF_UNSIGNED;
+                }
+                // LE_UN/GT_UN(expr, int/long.MaxValue) => GE/LT(expr, 0).
+                else if (((op1->TypeIs(TYP_LONG) && (op2Value == INT64_MAX))) ||
+                         ((genActualType(op1) == TYP_INT) && (op2Value == INT32_MAX)))
+                {
+                    oper = (oper == GT_LE) ? GT_GE : GT_LT;
+                    cmp->gtFlags &= ~GTF_UNSIGNED;
+                }
+            }
         }
-        // Check for "expr > -1".
-        else if (oper == GT_GT)
+
+        if (!cmp->OperIs(oper))
         {
-            // Change to "expr >= 0".
-            oper = GT_GE;
+            // Keep the old ValueNumber for 'tree' as the new expr
+            // will still compute the same value as before.
+            cmp->SetOper(oper, GenTree::PRESERVE_VN);
+            op2->SetIntegralValue(0);
+            fgUpdateConstTreeValueNumber(op2);
         }
     }
-    else if (cmp->IsUnsigned())
+
+    if (cmp->OperIs(GT_LT, GT_GT))
     {
-        if ((oper == GT_LE) || (oper == GT_GT))
+        assert(cmp->gtGetOp2()->IsIntegralConst() || cmp->gtGetOp1()->IsIntegralConst());
+        assert(!gtIsActiveCSE_Candidate(cmp->gtGetOp2()) || !gtIsActiveCSE_Candidate(cmp->gtGetOp1()));
+
+        GenTree*  conNode   = cmp->gtGetOp2()->IsIntegralConst() ? cmp->gtGetOp2() : cmp->gtGetOp1();
+        GenTree*  varNode   = cmp->gtGetOp2()->IsIntegralConst() ? cmp->gtGetOp1() : cmp->gtGetOp2();
+        var_types limitType = varNode->TypeGet();
+
+        if (varNode->OperIs(GT_CAST))
+        {
+            limitType = varNode->AsCast()->gtCastType;
+        }
+
+        switch (limitType)
         {
-            if (op2Value == 0)
+            case TYP_UBYTE:
+            case TYP_SHORT:
+            case TYP_USHORT:
+            case TYP_INT:
+            case TYP_UINT:
+            case TYP_LONG:
+                break;
+            default:
+                return cmp;
+        }
+
+        IntegralRange valRange = IntegralRange::ForNode(varNode, this);
+
+#if defined(HOST_X86) || defined(HOST_ARM)
+        ssize_t valMin = (int32_t)IntegralRange::SymbolicToRealValue(valRange.LowerBoundForType(limitType));
+        ssize_t valMax = (int32_t)IntegralRange::SymbolicToRealValue(valRange.UpperBoundForType(limitType));
+#else
+        ssize_t valMax      = IntegralRange::SymbolicToRealValue(valRange.UpperBoundForType(limitType));
+        ssize_t valMin      = IntegralRange::SymbolicToRealValue(valRange.LowerBoundForType(limitType));
+
+#endif
```

Comment:
Note that the `UpperBoundForType` and `LowerBoundForType` functions you are calling are static functions.
You should add methods like `IntegralRange::GetLowerBound() { return m_lowerBound; }` and `IntegralRange::GetUpperBound() { return m_upperBound; }` instead. Then all this logic should just use `IntegralRange::ForNode` and calls to `IntegralRange::SymbolicToRealValue(valRange.GetLowerBound())` and the same for the upper bound.